src/main/java/org/truffleruby/core/kernel/TruffleKernelNodes.java

```diff
@@ -92,6 +96,61 @@ public boolean load(VirtualFrame frame, DynamicObject file, boolean wrap,
+    public abstract static class SetGlobalVariableNode extends CoreMethodArrayArgumentsNode {
+        @Child protected WriteBarrierNode writeBarrierNode = WriteBarrierNode.create();
+
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName",
+                "referenceEqualNode.executeReferenceEqual(value, previousValue)" }, limit = "1", assumptions = "storage.getUnchangedAssumption()")
+        public Object writeTryToKeepConstant(DynamicObject name, Object value,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage,
+                @Cached("storage.getValue()") Object previousValue,
+                @Cached("create()") ReferenceEqualNode referenceEqualNode) {
+            // NOTE: we still do the volatile write to get the proper memory barrier,
+            // as the global variable could be used as a publication mechanism.
+            storage.setValueInternal(value);
+            return previousValue;
+        }
+
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName", "storage.isAssumeConstant()" }, limit = "1", assumptions = "storage.getUnchangedAssumption()")
+        public Object writeAssumeConstant(DynamicObject name, Object value,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage) {
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            storage.updateAssumeConstant(getContext());
+            return value;
+        }
+
+        @Specialization(guards = { "isRubySymbol(cachedName)", "name == cachedName", "!storage.isAssumeConstant()" }, limit = "1", replaces = "writeAssumeConstant")
+        public Object write(DynamicObject name, Object value,
+                @Cached("name") DynamicObject cachedName,
+                @Cached("getStorage(name)") GlobalVariableStorage storage) {
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            return value;
+        }
+
+        @Specialization(guards = "isRubySymbol(name)")
+        public Object writeGeneric(DynamicObject name, Object value) {
+            GlobalVariableStorage storage = getStorage(name);
+            if (getContext().getSharedObjects().isSharing()) {
+                writeBarrierNode.executeWriteBarrier(value);
+            }
+            storage.setValueInternal(value);
+            return value;
+        }
+
+        protected GlobalVariableStorage getStorage(DynamicObject name) {
+            return getContext().getCoreLibrary().getGlobalVariables().getStorage(Layouts.SYMBOL.getString(name));
+        }
+    }
```

Comment:
This looks like overkill for the simple `setValueInternal`.
And it duplicates a lot of complex code from elsewhere.